### ISSUE TITLE
Theme switching does not work with multiple desktops

### ISSUE THREAD

**Issue by ZhenmingYu** [AutoDarkMode/Windows-Auto-Night-Mode#469]
### Description

Theme switching does not work with multiple desktops.

For taskbar, start menu, and stuff, the app is working to automatically switch dark and light themes.
However, for the wallpaper, the app switching does not work, only the first virtual desktops among many desktops get switched to the correct wallpaper.

Manual switching with shortcut keys still works and could switch the current desktop to the correct theme.
However, the other desktops are still left behind with the old wallpaper from the other themes.

### Expected Behavior

ADM should be able to switch all desktop wallpapers together.

I guess this could be done with a for-loop, where the AMD switches all desktops sequentially.

### Log Data

[service.log](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/files/8910797/service.log)
[config.yaml.log](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/files/8910820/config.yaml.log)

### Operating System

Windows 11 pro 1000.22000.739.0

### Commit Hash and Version

- Commit: `5e6ea8a`
- Service: `10.1.0.10`
- Updater: `2.4.2`
- App: `10.0.1.0`
- Shell: `1.3.1.0`
- .Net: `6.0.0`

### Screenshots

<img width="996" alt="20220615172619" src="https://user-images.githubusercontent.com/8026796/173865882-5cd3764c-e9fe-4886-9034-f66ae61bbec7.png">

**Comment by Spiritreader** [comment]
duplicate of https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/issues/228

However, you've filled out the bug report template and provided screenshots, so I'm closing the other issue in favor of this one.

**Comment by ZhenmingYu** [comment]
Thanks,

This should be labeled as a "known issue" to make it easier to follow.

**Comment by Spiritreader** [comment]
It actually was at some point, dunno what happened to it.
I'll make this one the known issue

**Comment by JasonHK** [comment]
> Also, in theory the wallpaper switch should work fine if you use our Wallpaper switching mechanism instead of Windows theme mode, but it may also be broken in newer builds.

@Spiritreader Nope, wallpaper switch mode is broken as well

**Comment by JasonHK** [comment]
Just a idea. Is it possible to detect virtual desktop switching? If possible, maybe we can check which wallpaper/theme was used, and switch to the correct one when switched between virtual desktop?

**Comment by Spiritreader** [comment]
> > Also, in theory the wallpaper switch should work fine if you use our Wallpaper switching mechanism instead of Windows theme mode, but it may also be broken in newer builds.
> 
> @Spiritreader Nope, wallpaper switch mode is broken as well

@Armin2208 could you re-test this when you get the chance? Thai contradicts our internal testing.

**Comment by Armin2208** [comment]
> > Also, in theory the wallpaper switch should work fine if you use our Wallpaper switching mechanism instead of Windows theme mode, but it may also be broken in newer builds.
> 
> @Spiritreader Nope, wallpaper switch mode is broken as well

In which windows build are you facing issues with wallpaper switch?

**Comment by Spiritreader** [comment]
@JasonHK oh yes before I forget: did you use the multi monitor option when using our wallpaper switcher?
It won't work with the other one.

**Comment by JasonHK** [comment]
> > > Also, in theory the wallpaper switch should work fine if you use our Wallpaper switching mechanism instead of Windows theme mode, but it may also be broken in newer builds.
> > 
> > 
> > @Spiritreader Nope, wallpaper switch mode is broken as well
> 
> In which windows build are you facing issues with wallpaper switch?

@Armin2208 22000.856

> @JasonHK oh yes before I forget: did you use the multi monitor option when using our wallpaper switcher? It won't work with the other one. You have to use the multi monitor one as this should toggle a refresh of all vdesktop wallpapers.

@Spiritreader That one doesn't work either

**Comment by carlosvsilva** [comment]
I have started to notice this major bug only after upgrade from Windows 11 21H2 to 22H2 latest Release Preview. Switching to Dark Theme is nearly instantaneous, but latest Windows Version 10.0.22621.382 forgets to change the wallpaper on all other virtual desktops, other than the current one in use.

I've reported the Windows bug (happens even when switching from Windows Settings) to Microsoft which breaks theming, and Auto Dark Mode, but the September 20th release date is fast approaching and the bug report remains unacknowledged (?).
The only way to workaround this, is by going to every individual virtual desktop, open Windows Settings and click Personalization. Only then Windows changes the wallpaper to match the current theme.

If you want to help upvote the bug report, it's here:

https://aka.ms/AAhmoj9

Thanks!

**Comment by AlbertoFabbri93** [comment]
When I try to open your link with the Feedback Hub I get the following error:
"Your account doesn´t have access to this feedback."

**Comment by carlosvsilva** [comment]
I shared the feedback via the app "feedback hub" which produced that short URL, if it's forbidden, must be some problem on Microsoft side.

**Comment by Spiritreader** [comment]
The proper way to get ADM's own wallpaper switcher to work is this

- Close all open virtual desktops (extremely important!!!)
- Open Auto Dark Mode App
- Go to the Personalization page
- Click on Pick a wallpaper and enable it
- Select light select Type: Picture (multiple monitors). Doesn't matter if you only have one monitor, select that. Then pick your wallpaper
- Select darkand also select Type: Picture (multiple monitors). Then pick your dark wallpaper
- Reopen your virtual desktops
- Wallpapers will now switch

If that doesn't work, then you're not using the multi-monitor switching method that works with vdesktops.

Support for V-Desktops will come at some point, but there's still no package yet that provides this functionality with version-specific implementation. Until that is available, the method described above is the only way to get it working withour wallpaper switcher. The alternatives open to you are using the Windows theme mode, or not using the wallpaper switcher / using a third party wallpaper tool.
They will start to work fine with ADM 10.3, which is already in beta and should be pretty stable as we're preparing to push it to stable.

**Comment by jczimm** [comment]
Using version 10.3 on Windows 11 and neither method @Spiritreader shared worked for me (the Picture (multiple monitors) method in bullets or using the Windows theme mode).

However, I thought I'd share a different workaround I came up with. I enabled the Windows theme mode and set it to toggle between two themes that each use a Slideshow (instead of a Picture), and I have it do a slideshow over a folder with just one image, configured to change once per day. (Windows should rarely try to cycle the slideshow; if I miss a theme cycle (e.g. entire day or entire night), though I think there might just be a short moment of lag when the theme has been up for a whole day and the theme reloads the image as the "next" image in the slideshow). Think this should work great for me. Also, I get the added benefit of the wallpaper transitioning when theme changes now!

**Comment by HoeYeungHo** [comment]
I am not sure whether the above operation can work. However, if I choose to switch themes automatically. I cannot pick a wallpaper with warnning: This feature is disabled while theme mode is enabled.
I prefer that theme colors and wallpapers can be synchronized on multiple desktops.

ADM version:10.1.0.10
Windows version: Windows 11 22H2

**Comment by Spiritreader** [comment]
> I am not sure whether the above operation can work. However, if I choose to switch themes automatically. I cannot pick a wallpaper with warnning: This feature is disabled while theme mode is enabled. I prefer that theme colors and wallpapers can be synchronized on multiple desktops.
> 
> ADM version:10.1.0.10 Windows version: Windows 11 22H2

Yes that is normal.
You can either switch between two different themes, or let auto dark Mode modify one single theme.

If you use the theme mode, everything is managed by windows. If windows is buggy, then you'll have to live with it basically.
You can manually create themes that use multi monitor wallpaper switching if you want. Check out Microsofts documentation on how to create theme files and that should do the trick.

**Comment by ShunbaoLi** [comment]
I am having the same issue.

But the situation on my side is a little bit different.

I found that the wallpaper will still switch automatically, but there will be a delay.

If you stay on the desktop whose wallpaper was not automatically switched for about a minute, the wallpaper will still auto-switch as expected.

**Comment by Spiritreader** [comment]
Not stale

**Comment by RouisIbra** [comment]
@Spiritreader 
> Support for V-Desktops will come at some point, but there's still no package yet that provides this functionality with version-specific implementation.

There is, It is a Powershell module [PSVirtualDesktop](https://github.com/MScholtes/PSVirtualDesktop) and it can set wallpaper on all virtual desktops like this:

    Set-AllDesktopWallpapers -Path path

**Comment by AlesSrsen** [comment]
[Windows Auto Dark Mode - Change multi-screen wallpaper](https://valh.io/p/windows-auto-dark-mode-change-multi-screen-wallpaper/)
I've come across this article while searching for this issue, do you think this might be at least a temporary solution? _Disclaimer, I haven't tried it since I want to use the switching of the whole theme. (But this might be useful also for that)_

**Comment by Spiritreader** [comment]
> [Windows Auto Dark Mode - Change multi-screen wallpaper](https://valh.io/p/windows-auto-dark-mode-change-multi-screen-wallpaper/) I've come across this article while searching for this issue, do you think this might be at least a temporary solution? _Disclaimer, I haven't tried it since I want to use the switching of the whole theme. (But this might be useful also for that)_

Hey!
No this doesn't apply.
The article itself is only half-correct information in the first place, since Auto Dark Mode does support multi screen wallpapers, just not the Span argument.

**Comment by ThioJoe** [comment]
> @Spiritreader
> 
> > Support for V-Desktops will come at some point, but there's still no package yet that provides this functionality with version-specific implementation.
> 
> There is, It is a Powershell module [PSVirtualDesktop](https://github.com/MScholtes/PSVirtualDesktop) and it can set wallpaper on all virtual desktops like this:
> 
> ```
> Set-AllDesktopWallpapers -Path path
> ```
> 
> It has also a C# version with version-specific implementation: https://github.com/MScholtes/VirtualDesktop

I was able to get everything working with virtual desktops using that powershell module and using the custom scripts feature. I just enabled custom scripts and disabled the "Enable wallpaper switch" option so everything is controlled individually with the script. Here's an example of my yaml:

```
Enabled: true
Component:
  Scripts:
  - Name: Set Wallpaper Light Mode
    Command: powershell
    ArgsLight:
      - -Command
      - "Import-Module VirtualDesktop; Set-DesktopWallpaper -Desktop 'WhateverDesktopName1' -Path 'C:\\WhateverPath\\LightWallpaper.jpg' -PassThru; Set-DesktopWallpaper -Desktop 'WhateverDesktopName2' -Path 'C:\\WhateverPath\\LightWallpaper.jpg' -PassThru"
    ArgsDark: []
    AllowedSources: [Any]
    
  - Name: Set Wallpaper Dark Mode
    Command: powershell
    ArgsLight: []
    ArgsDark:
      - -Command
      - "Import-Module VirtualDesktop; Set-DesktopWallpaper -Desktop 'WhateverDesktopName1' -Path 'C:\\WhateverPath\\DarkWallpaper.jpg' -PassThru; Set-DesktopWallpaper -Desktop 'WhateverDesktopName2' -Path 'C:\\WhateverPath\\DarkWallpaper.jpg' -PassThru"
    AllowedSources: [Any]
```

Where you'd replace your desktop names instead of WhateverDesktopName and the wallpaper paths. That one is an example for two virtual desktops, but you can add more by doing another semicolon and copying the same Set-DesktopWallpaper command a third time.

Note that right now the module has an issue with the latest Windows 11 builds so at the moment you need to use a special version of the script the person made as described here:
https://github.com/MScholtes/PSVirtualDesktop/issues/37#issuecomment-1927273764

**Comment by VapidLinus** [comment]
Span mode does not work with dark-mode switching for me.

- I am in light mode
- I switch to dark mode
- Dark mode activates like it should
- I switch to light mode
- Light mode works briefly
- The dark wallpaper activates even though I am in light mode

Is this the same issue as here, or should I report this separately? And just to be clear, I recorded a video of what it looks like:

https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/assets/4418470/a0d72d4e-4a21-4427-97b2-8e8aaf09905f

Mirror: https://files.vapid.li/24/vuY5FsCVsw.mp4

**These are my settings:**
![AutoDarkModeApp_IOc9lO4VHy](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/assets/4418470/a194d02b-b4eb-4d6a-9046-54b768902e36)
![AutoDarkModeApp_7drDhl6fd9](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/assets/4418470/d0ee3983-bd7b-43a6-9e5b-3817c13b0fd0)
![ApplicationFrameHost_zk2db01hxj](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/assets/4418470/1804f477-0ac6-4b83-9433-ed5297bcaf77)

**Comment by Spiritreader** [comment]
> Span mode does not work with dark-mode switching for me.
> 
>     * I am in light mode
> 
>     * I switch to dark mode
> 
>     * Dark mode activates like it should
> 
>     * I switch to light mode
> 
>     * Light mode works briefly
> 
>     * The dark wallpaper activates even though I am in light mode
> 
> 
> **Is this the same issue as here, or should I report this separately?** And just to be clear, I recorded a video of what it looks like:
> vuY5FsCVsw.mp4
> 
> Mirror: https://files.vapid.li/24/vuY5FsCVsw.mp4
> 
> **These are my settings:** ![AutoDarkModeApp_IOc9lO4VHy](https://private-user-images.githubusercontent.com/4418470/311496170-a194d02b-b4eb-4d6a-9046-54b768902e36.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwNjM2MDMsIm5iZiI6MTcxMDA2MzMwMywicGF0aCI6Ii80NDE4NDcwLzMxMTQ5NjE3MC1hMTk0ZDAyYi1iNGViLTRkNmEtOTA0Ni01NGI3Njg5MDJlMzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTBUMDkzNTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDM2ZTYyNTE2M2UxZWJmNzQwYjkyODZkZDdiMmJlY2RhZjczMjU1MmY0N2M4YTg5ZGEyOTZiNDMzMWM5Mzk4YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.MXoNvFvrb1N6GJTvMgZWsrGe_nsS08I7VDT1yl1fzA8) ![AutoDarkModeApp_7drDhl6fd9](https://private-user-images.githubusercontent.com/4418470/311496174-d0ee3983-bd7b-43a6-9e5b-3817c13b0fd0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwNjM2MDMsIm5iZiI6MTcxMDA2MzMwMywicGF0aCI6Ii80NDE4NDcwLzMxMTQ5NjE3NC1kMGVlMzk4My1iZDdiLTQzYTYtOWU1Yi0zODE3YzEzYjBmZDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTBUMDkzNTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDk5ODQzYWQ5YWRjMzMzNWI3YTU0YTFjYzYwYmUwNjFiMThiMDNjZmM4MmRiZmQ0MWI1MTljNDhjMTE0ODJjMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.n9wJ7i_FBVn5zG2KC9Oc1dAowF_Mb9iWgQoLnCl5yiY) ![ApplicationFrameHost_zk2db01hxj](https://private-user-images.githubusercontent.com/4418470/311496176-1804f477-0ac6-4b83-9433-ed5297bcaf77.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAwNjM2MDMsIm5iZiI6MTcxMDA2MzMwMywicGF0aCI6Ii80NDE4NDcwLzMxMTQ5NjE3Ni0xODA0ZjQ3Ny0wYWM2LTRiODMtOTQzMy1lZDUyOTdiY2FmNzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTBUMDkzNTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjlhMzEwYTkxMDA4ZDNkNGExYzVhYzM0ZWMzNTFiZWQ4N2I3MDhiZDBjYzI2OGZkYWRiMTRiMGIyNGMxNDZkNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nOd4X7WV3cMXgkZDJMRR9K5R0wtyrAsadVYYqSWEOiU)

That sounds like you aren't using the latest build.
Check your settings and verify you're using the latest beta build (which is up to date with the master branch)

**Comment by VapidLinus** [comment]
@Spiritreader I switched to the beta branch, and it did indeed fix the problem. Thank you!

**Comment by themooleman** [comment]
> The proper way to get ADM's own wallpaper switcher to work is this
> 
>     * Close all open virtual desktops (extremely important!!!)
> 
>     * Open Auto Dark Mode App
> 
>     * Go to the Personalization page
> 
>     * Click on Pick a wallpaper and enable it
> 
>     * Select Light and select Type: Picture (multiple monitors). Doesn't matter if you only have one monitor, select that. Then pick your wallpaper
> 
>     * Select Dark and also select Type: Picture (multiple monitors). Then pick your dark wallpaper
> 
>     * Reopen your virtual desktops
> 
>     * Wallpapers will now switch
> 
> 
> If that doesn't work, then you're still not using the multi-monitor switching method that works with vdesktops. Try again starting with step 1.
> 
> Support for V-Desktops will come at some point, but there's still no package yet that provides this functionality with version-specific implementation. Until that is available, the method described above is the only way to get it working with our wallpaper switcher. The alternatives open to you are using the Windows theme mode, or not using the wallpaper switcher / using a third party wallpaper tool. They will start to work fine with ADM 10.3, which is already in beta and should be pretty stable as we're preparing to push it to stable.

This worked a treat for me. Thanks for the explicit, step-by-step guide.

**Comment by SteynGuelen** [comment]
> The proper way to get ADM's own wallpaper switcher to work is this
> 
> * Close all open virtual desktops (extremely important!!!)
> * Open Auto Dark Mode App
> * Go to the Personalization page
> * Click on Pick a wallpaper and enable it
> * Select Light and select Type: Picture (multiple monitors). Doesn't matter if you only have one monitor, select that. Then pick your wallpaper
> * Select Dark and also select Type: Picture (multiple monitors). Then pick your dark wallpaper
> * Reopen your virtual desktops
> * Wallpapers will now switch
> 
> If that doesn't work, then you're still not using the multi-monitor switching method that works with vdesktops. Try again starting with step 1.
> 
> Support for V-Desktops will come at some point, but there's still no package yet that provides this functionality with version-specific implementation. Until that is available, the method described above is the only way to get it working with our wallpaper switcher. The alternatives open to you are using the Windows theme mode, or not using the wallpaper switcher / using a third party wallpaper tool. They will start to work fine with ADM 10.3, which is already in beta and should be pretty stable as we're preparing to push it to stable.

Can confirm that this works, and is so far the only solution that keeps on working. Thanks for the guide!

**Comment by Spiritreader** [comment]
not a hardware issue, and still relevant. This should stay pinned until we have a permanent solution.

**Comment by JohnRenzFlores** [comment]
Is there any update on this feature?

**Comment by Armin2208** [comment]
> Is there any update on this feature?

Nope, nothing changed.
